device/send.go

```diff
@@ -212,81 +213,124 @@ func (device *Device) RoutineReadFromTUN() {
 
-	var elem *QueueOutboundElement
+	var (
+		readErr     error
+		elems       = make([]*QueueOutboundElement, device.BatchSize())
+		buffs       = make([][]byte, device.BatchSize())
+		elemsByPeer = make(map[*Peer]*[]*QueueOutboundElement, device.BatchSize())
+		count       = 0
+		sizes       = make([]int, device.BatchSize())
+		offset      = MessageTransportHeaderSize
+	)
 
-	for {
-		if elem != nil {
-			device.PutMessageBuffer(elem.buffer)
-			device.PutOutboundElement(elem)
+	defer func() {
+		for _, elem := range elems {
+			if elem != nil {
+				device.PutMessageBuffer(elem.buffer)
+				device.PutOutboundElement(elem)
+			}
 		}
-		elem = device.NewOutboundElement()
-
-		// read packet
+	}()
 
-		offset := MessageTransportHeaderSize
-		size, err := device.tun.device.Read(elem.buffer[:], offset)
-		if err != nil {
-			if !device.isClosed() {
-				if !errors.Is(err, os.ErrClosed) {
-					device.log.Errorf("Failed to read packet from TUN device: %v", err)
-				}
-				go device.Close()
+	for {
+		// Repopulate queue elements and buffer pointers
+		for i := range elems {
+			if elems[i] == nil {
+				elems[i] = device.NewOutboundElement()
 			}
-			device.PutMessageBuffer(elem.buffer)
-			device.PutOutboundElement(elem)
-			return
+			buffs[i] = elems[i].buffer[:]
 		}
 
-		if size == 0 || size > MaxContentSize {
-			continue
-		}
+		// read packets
+		count, readErr = device.tun.device.Read(buffs, sizes, offset)
+		for i := 0; i < count; i++ {
+			if sizes[i] < 1 {
+				continue
+			}
 
-		elem.packet = elem.buffer[offset : offset+size]
+			elem := elems[i]
+			elem.packet = buffs[i][offset : offset+sizes[i]]
 
-		// lookup peer
+			// lookup peer
+			var peer *Peer
+			switch elem.packet[0] >> 4 {
+			case ipv4.Version:
+				if len(elem.packet) < ipv4.HeaderLen {
+					continue
+				}
+				dst := elem.packet[IPv4offsetDst : IPv4offsetDst+net.IPv4len]
+				peer = device.allowedips.Lookup(dst)
 
-		var peer *Peer
-		switch elem.packet[0] >> 4 {
-		case ipv4.Version:
-			if len(elem.packet) < ipv4.HeaderLen {
-				continue
+			case ipv6.Version:
+				if len(elem.packet) < ipv6.HeaderLen {
+					continue
+				}
+				dst := elem.packet[IPv6offsetDst : IPv6offsetDst+net.IPv6len]
+				peer = device.allowedips.Lookup(dst)
+
+			default:
+				device.log.Verbosef("Received packet with unknown IP version")
 			}
-			dst := elem.packet[IPv4offsetDst : IPv4offsetDst+net.IPv4len]
-			peer = device.allowedips.Lookup(dst)
 
-		case ipv6.Version:
-			if len(elem.packet) < ipv6.HeaderLen {
+			if peer == nil {
 				continue
 			}
-			dst := elem.packet[IPv6offsetDst : IPv6offsetDst+net.IPv6len]
-			peer = device.allowedips.Lookup(dst)
-
-		default:
-			device.log.Verbosef("Received packet with unknown IP version")
+			elemsForPeer, ok := elemsByPeer[peer]
+			if !ok {
+				elemsForPeer = device.GetOutboundElementsSlice()
+				elemsByPeer[peer] = elemsForPeer
+			}
+			*elemsForPeer = append(*elemsForPeer, elem)
+			elems[i] = nil
```

Comment:
The way this function worked before is that it would only need allocate a new elem if one was sent. If the peer wasn't yet running, then it would just reuse that elem on the next iteration.

Now it's more complex because of this shuffling into the elemsByPeer hash table. However, in the process, parts of the original design were kept in modified form. Namely, at the top of each iteration, all null elems are refilled, by iterating through the whole list. I wonder: would it be better to instead refill it when nil'ing it out on this line here? Instead of `elems[i] = nil`, you could do `elems[i] = device.NewOutboundElement(), bufs[i] = ..`? This would avoid that iteration at the top of the loop; the initial filling could be moved to the function preamble. Or do you see advantages to doing it how it's done now.

Reply:
Avoiding iterating the whole list is better. Updated in https://github.com/WireGuard/wireguard-go/pull/64/commits/52e0653b682325de302a4ef4ee8537aa664fb590.